### DISCUSSION TITLE
How to Write and Deploy Your First Smart Contract on Soroban

### DISCUSSION THREAD

**Discussion by AdityaMisra01** [stellar/sorobounty-spectacular#10] [General]
## Introduction

A smart contract is code that is deployed and executed on a blockchain. Soroban allows writing smart contracts using the Rust programming language.

In this tutorial, we will:

1. Set up our development environment
2. Write a simple smart contract 
3. Compile the Rust code to WASM
4. Deploy the contract to Stellar's testnet
5. Interact with the deployed contract

By the end, you will have hands-on experience deploying a decentralized application on Soroban from start to finish.

## Set Up the Development Environment

First, we need to install the necessary tools:

- **Rust** - The language Soroban contracts are written in

  - Install via https://rustup.rs
  - Check installation with `rustc --version`

- **Soroban CLI** - To interact locally with Soroban contracts

  - Install via https://soroban.stellar.org/docs/category/getting-started 
  - Check with `soroban --version`

- **Git** - For initializing projects and source control

  - Install based on your operating system
  - Verify with `git --version` 

- **Futurenet** - Stellar's Soroban test network

  - Create account at https://laboratory.stellar.org
  - Fund account with test XLM

- **Text Editor** - VS Code recommended for writing code

  - Install from https://code.visualstudio.com
  - Useful Rust plugins: Rust Analyzer, Better TOML

With these installed, we are ready to start developing!

## Write the Smart Contract

Now we can write our first contract. 

- Create a new Rust project:

```
cargo new --lib soroban-tutorial
```

- Open the main `src/lib.rs` file, we will write our contract here:

```rust
use soroban_sdk::*;

#[sc_contract]
pub struct Counter {
  count: i64,
}

impl Counter {

  pub fn new() -> Self {
    Self { count: 0 }
  }

  pub fn increment(&mut self) {
    self.count += 1;
  }

  pub fn get_count(&self) -> i64 {
    self.count
  }

}
```

This implements a simple counter variable that can be incremented.

- The `#[sc_contract]` macro defines the struct that stores state
- `new()` is the constructor that initializes state
- `increment()` modifies state
- `get_count()` reads state

Now let's compile this contract...

## Compile the Contract 

To deploy to Soroban, we need to compile the Rust code to WASM bytecode.

- Build the release version:

```
cargo build --release
```

- This creates a WASM file in `target/release`. 

- We can optimize the bytecode size further:

```
wasm-opt -Oz target/release/soroban_tutorial.wasm -o target/optimized.wasm
```

- Verify it works locally using:

```
cargo test 
```

Now we have an optimized WASM contract ready to deploy!

## Deploy the Contract

With our compiled contract, deploying to Soroban's Futurenet takes just a few commands:

- First, make sure the Soroban CLI is configured properly:

```
soroban wallet import $YOUR_ACCOUNT_KEY --testnet futurenet
```

- Then deploy the contract: 

```
soroban contract deploy target/optimized.wasm --testnet futurenet
```

- This will return the contract ID. Save it for later use.

- Fund the contract's public key with a few test XLM.

- View details using:

```
soroban contract get $CONTRACT_ID --testnet futurenet
```

Our contract is now live on Futurenet!

## Interact with the Contract

We can call methods on our deployed contract using the CLI:

```
// Increment
soroban contract execute $CONTRACT_ID increment --testnet futurenet

// Get current count 
soroban contract query $CONTRACT_ID get_count --testnet futurenet
```

The contract functions exactly as we coded it!

## Recap and Next Steps

In this tutorial, we:

- Set up our local dev environment
- Wrote a simple smart contract in Rust 
- Compiled the code to optimized WASM bytecode
- Deployed it to Stellar's Soroban testnet  
- Interacted with the contract methods

For next steps, you could:

- Expand the contract with more features
- Add a frontend UI for easier interaction
- Deploy to mainnet for production use

With the basics covered here, you are ready to start building more complex decentralized apps on Soroban!

**Comment by AdityaMisra01** [comment]
This link is broken 
https://docs.soroban.stellar.org/docs/getting-started

**Reply by rice2000** [reply]
Is this the link you're looking for? https://soroban.stellar.org/docs/category/getting-started

The one you shared doesn't look quite right because there is no docs.soroban.stellar.org.  There's only soroban.stellar.org/docs

**Reply by AdityaMisra01** [reply]
Yes, Thank you I'll update the link in the tutorial.
Will this tutorial be eligible for the reward?

**Comment by briwylde08** [comment]
Hi there! Thank you for your submission. Unfortunately, this doesn’t quite follow the eligibility guidelines for the Sorobounty Spectacular program but you’re welcome to edit your submission to be eligible!

The minimum requirement for a submission is dapp code that is publicly available in a GitHub repo (we have some great dapp ideas published [here](https://github.com/stellar/sorobounty-spectacular/blob/main/RFPs.md)) and then you can add on educational content (such as written tutorials and video walkthroughs) published in a GitHub README, [dev.to](http://dev.to/) or Medium article, or YouTube video.

Please let me know if you have any questions!

**Comment by AdityaMisra01** [comment]
Okay, thank you!
